File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
def solution(numbers):
    answer = 0
    candidate = set()
    
    def backtracking(cur, t):
        if len(cur) == t:
            val = ''
            for c in cur:
                val += numbers[c]
            candidate.add(int(val))
            return
        
        for idx in range(len(numbers)):
            if idx not in cur:
                cur.append(idx)
                backtracking(cur, t)
                cur.pop()
    
    for target in range(1, len(numbers)+1):
        backtracking([], target)
    
    def isPrime(n):
        if n<2: return False
        for i in range(2, int(n**0.5)+1):
            if n%i == 0:
                return False
        return True
    
    for c in candidate:
        if isPrime(c): 
            answer += 1
    
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <functional>
#include <set>
#include <iostream>

using namespace std;

int solution(string numbers) {
    int answer = 0;
    set<int> candidate;
    
    function<bool(int)> is_prime = [](int target) {
        if(target < 2) return false;
        for(int i=2;i*i<=target;i++) {
            if (target%i == 0) return false;
        }
        return true;
    };
    
    function<void(vector<int>)> backtracking = [&](vector<int> curr) {
        if (!curr.empty()) {
            string tmp = "";
            for (auto idx: curr) {
                tmp += numbers[idx];
            }
            candidate.insert(stoi(tmp));
        }
        
        for(int i=0;i<numbers.length();i++) {
            if (find(curr.begin(), curr.end(), i) == curr.end()) {
                curr.push_back(i);
                backtracking(curr);
                curr.pop_back();
            }
        }
    };
    backtracking(vector<int>());
    for(auto c: candidate) {
        if (is_prime(c)) {
            answer++;
        }
    }
    
    return answer;
}